storage/innobase/lock/lock0lock.cc

```diff
@@ -3976,6 +3978,36 @@ dberr_t lock_table_for_trx(dict_table_t *table, trx_t *trx, lock_mode mode,
   return err;
 }
 
+/** Lock the child tables of a table.
```

Comment:
I agree that it is a mess, and I have long advocated the view that the SQL layer should acquire necessary and sufficient MDL on all DML and DDL operations. We also lack an API for debug checks whether a MDL is being held on a specific name.

It could be a good exercise to compare [WL#6049](https://dev.mysql.com/worklog/task/?id=6049) and the MariaDB implementation. [MDEV-22180](https://jira.mariadb.org/browse/MDEV-22180) seems to be a worthwhile starting point; it also links to the still open bug [MDEV-30103](https://jira.mariadb.org/browse/MDEV-30103). Starting with [MDEV-21175](https://jira.mariadb.org/browse/MDEV-21175) in MariaDB Server 10.6 we have no work-arounds for insufficient MDL during DDL operations.

There is also [MDEV-26990](https://jira.mariadb.org/browse/MDEV-26990) for an easily repeatable race condition between `INSERT` and `RENAME TABLE`, which I understood that @vuvova agreed to fix by making the SQL layer acquire more MDL, on DML, not DDL. He seems to be of the view that because InnoDB is storing the `FOREIGN KEY` relationships, it is InnoDB’s responsibility to acquire sufficient locks in those DDL operations that involve updating the relationships in the data dictionary or the cache objects of parent or child tables (including cases where the constraints themselves do not change).

Reply:
Any `FOREIGN KEY` trouble will be checked by the SQL layer before the execution reaches `ha_innobase::truncate()`, so we can safely avoid the waits in `lock_table_children()` when `foreign_key_checks=1`.

 `RENAME TABLE` is different: the following would enter `lock_wait()` in `lock_table_for_trx()` inside `ha_innobase::rename_table()`:
 ```diff
 diff --git a/mysql-test/suite/innodb/t/foreign_key.test b/mysql-test/suite/innodb/t/foreign_key.test
index 0db3a7ca377..623f0ec6a46 100644
--- a/mysql-test/suite/innodb/t/foreign_key.test
+++ b/mysql-test/suite/innodb/t/foreign_key.test
@@ -1077,6 +1077,9 @@ BEGIN;
 --error ER_NO_REFERENCED_ROW_2
 INSERT INTO child SET a=1;
 connection default;
+--error ER_TRUNCATE_ILLEGAL_FK
+TRUNCATE TABLE parent;
+RENAME TABLE parent TO transparent;
 SET innodb_lock_wait_timeout=0, foreign_key_checks=0;
 --error ER_LOCK_WAIT_TIMEOUT
 TRUNCATE TABLE parent;
```
If I change the `RENAME` statement to `ALTER TABLE parent FORCE, ALGORITHM=COPY` then it would get lock wait timeout at the same spot, `ha_innobase::rename_table()` for renaming `parent` to `#sql-backup-`something.

If I change the statement to `ALTER TABLE parent FORCE, ALGORITHM=INPLACE` then the `lock_wait_for_trx()` in `ha_innobase::commit_inplace_alter_table()` would time out.

There is one more `lock_table_children()` caller: `ha_innobase::delete_table()`. The statement `DROP TABLE parent`  would time out in `lock_table_for_trx()`.

I will adjust all these functions `ha_innobase::rename_table()`, `ha_innobase::commit_inplace_alter_table()`, `ha_innobase::delete_table()` so that they will avoid the lock wait timeout in case an error would be flagged due to `foreign_key_checks=1`.

Reply:
Because of the following test in `innodb.innodb-truncate` for MDEV-18923, we’d better not change the logic of `ha_innobase::rename_table()` to flag errors before locking:
```sql
call mtr.add_suppression('InnoDB: in RENAME TABLE table `test`.`t3`');
SET FOREIGN_KEY_CHECKS= OFF;
CREATE TABLE t1 (f2 INT, f4 INT, KEY(f2), FOREIGN KEY (f4) REFERENCES t3 (f4)) ENGINE=InnoDB;
SET FOREIGN_KEY_CHECKS= ON;
CREATE TABLE t2 (f2 INT, FOREIGN KEY(f2) REFERENCES t1 (f2)) ENGINE=InnoDB;
--error ER_CANT_CREATE_TABLE
CREATE TABLE t3 (a INT) ENGINE=InnoDB;
ALTER TABLE t1 RENAME TO t3;
```
The renaming from `t1` to `t3` is expected to succeed. It is correcting the inconsistent `FOREIGN KEY` definition into a consistent one. We can only optimize the locking in `TRUNCATE` and `DROP TABLE`.